keras_nlp/models/gpt-neox/gpt_neox_attention.py

```diff
@@ -0,0 +1,102 @@
+import tensorflow as tf
+from tensorflow import keras
+from keras_nlp.layers.transformer_layer_utils import compute_causal_mask
+from rotary_embedding import RotaryEmbedding
+
+class GPTNeoXAttention(keras.layers.Layer):
+
+  def __init__(self, 
+               num_heads, 
+               hidden_dim, 
+               rotary_pct=0.25, 
+               max_position_embeddings=2048):
+    
+    super().__init__()
+    self.num_heads = num_heads
+    self.hidden_dim = hidden_dim
+    self.head_dim = hidden_dim // num_heads
+    self.rotary_dim = self.head_dim * rotary_pct
+    self.max_position_embeddings = max_position_embeddings
+    self.rotary_embedding = RotaryEmbedding(self.rotary_pct)
+    self.qkv = keras.layers.Dense(3 * self.hidden_dim) 
+    self.dense = keras.layers.Dense(self.hidden_dim)
+
+  def _compute_attention(self, 
+                         query, 
+                         key, 
+                         value, 
+                         attention_mask=None,
+                         head_mask=None):
+    
+    batch_size, _, query_len, _ = tf.shape(query)
+    key_len = tf.shape(key)[-2]
+    # causal_mask = self.bias[:, :, key_len - query_len : key_len, :key_len]
+    causal_mask = compute_causal_mask(batch_size, key_len, key_len)
+    
+    query = tf.reshape(query, [batch_size * self.num_heads, query_len, self.head_dim])
+    key = tf.reshape(key, [batch_size * self.num_heads, query_len, self.head_dim])
+    attention_scores = tf.zeros(
+        [batch_size * self.num_heads, query_len, self.head_dim],
+        dtype=query.dtype
+    )
+
+    attention_scores = tf.linalg.matmul(
+        attention_scores,
+        query,
+        tf.transpose(key, perm=[0, 2, 1]),
+        beta=1.0,
+        alpha=(tf.constant(1.0))
+    )
+    attention_scores = tf.reshape(attention_scores,
+                                  [batch_size, self.num_heads, query_len, key_len])
+    mask_value = tf.constant(float('-inf'), dtype=attention_scores.dtype)
+    attention_scores = tf.where(causal_mask, attention_scores, mask_value)
+
+    if attention_mask is not None:
+      attention_scores += attention_mask
+
+    attention_scores = tf.cast(tf.nn.softmax(attention_scores, axis=-1), dtype=value.dtype)
+    
+    if head_mask is not None:
+      attention_scores *= head_mask
+
+    attention_output = tf.matmul(attention_scores, value)
+    return attention_output, attention_scores
+
+
+  def call(self, 
```

Comment:
**Blocker**

I'm still confused about interfacing between GPTNeoXAttention and GPTNeoXDecoder with respect to call arguments (hidden_states, attention_mask, head_mask ), therefore I need your help here